Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentEnd/CookieCutterHelper.js

```diff
@@ -1,41 +1,73 @@
+    cookiePreferences["analytic"] = preferences["analytic"];
+    cookiePreferences["social"] = preferences["social"];
+}
 
-// Tell the iOS app to increase the count of cookies handled.
-CookieEngine.incrementCookieStats(async () => {
-    webkit.messageHandlers.cookieCutterHandler.postMessage({ update: "increase cookie stats"});
+Object.defineProperty(window.__firefox__, "setPreference", {
+    enumerable: false,
+    configurable: false,
+    writable: false,
+    value: setPreference
 });
 
-// Tell the iOS app that a cookie notice has been handled.
-CookieEngine.notifyNoticeHandledOnPage(async () => {
-    webkit.messageHandlers.cookieCutterHandler.postMessage({ update: "cookie notice handled"});
-});
+// Checks if the Cookie Cutter handler has been injected by iOS.
+// Without it, could cause other scripts to fail.
+if (webkit.messageHandlers.cookieCutterHandler != undefined) {
+    webkit.messageHandlers.cookieCutterHandler.postMessage({ update: "get-preferences"});
 
-// Needed if the page is reloaded.
-CookieEngine.getHostname(() => window.location);
+    // Not used by iOS.
+    CookieEngine.flagSite(async () => {});
 
-//
-// User preferences, passed down from the iOS app.
-// Set to true for testing.
-// TODO: Pass values from iOS app, and return them here.
-CookieEngine.areAllEnabled(async () => true);
-CookieEngine.isTypeEnabled(async (type) => true);
+    // These are handled by the iOS app, can return default values.
+    CookieEngine.isCookieConsentingEnabled(async () => true);
+    CookieEngine.isFlaggedSite(async () => false);
 
-//
-// TODO: Logging
-CookieEngine.logProviderUsage(async (provider) => {});
+    // Tell the iOS app to increase the count of cookies handled.
+    CookieEngine.incrementCookieStats(async () => {
+        webkit.messageHandlers.cookieCutterHandler.postMessage({ update: "increase-cookie-stats"});
+    });
 
-// Run!
-CookieEngine.runCookieCutter();
+    // Tell the iOS app that a cookie notice has been handled.
+    CookieEngine.notifyNoticeHandledOnPage(async () => {
+        webkit.messageHandlers.cookieCutterHandler.postMessage({ update: "cookie-notice-handled"});
+    });
 
-if (webkit.messageHandlers.cookieCutterHandler != undefined) {
-    webkit.messageHandlers.cookieCutterHandler.postMessage({ update: "started running"});
-}
\ No newline at end of file
+    // Needed if the page is reloaded.
+    CookieEngine.getHostname(() => window.location);
+
+    //
+    // User preferences, passed down from the iOS app.
+    CookieEngine.areAllEnabled(async () => {
+        return cookiePreferences["marketing"] && cookiePreferences["analytic"] && cookiePreferences["social"];
+    });
+
+    CookieEngine.isTypeEnabled(async (type) => {
+        switch (type) {
+        case 2:
+            return cookiePreferences.marketing;
+        case 3:
+            return cookiePreferences.analytic;
+        case 4:
+            return cookiePreferences.social;
+        default:
+            return true;
+        }
```

Comment:
The engine categories are actually a little more granular, but we can switch over them and just generalize to the three categories we set user preferences for. Also probably a good idea to decline cookies if our preferences don't explicitly allow them.
```suggestion
        switch (type) {
        case CookieCategoryType.Marketing:
        case CookieCategoryType.DoNotSell:
            return cookiePreferences.marketing;
        case CookieCategoryType.Analytics:
        case CookieCategoryType.Preferences:
            return cookiePreferences.analytic;
        case CookieCategoryType.Social:
        case CookieCategoryType.Unknown:
            return cookiePreferences.social;
        default:
            return false;
        }
```